### PR TITLE
add all binary morphology functions to cupyx.scipy.ndimage

### PR DESCRIPTION
**Pull request by grlee77**
This PR adds `binary_erosion`, `binary_dilation`, `binary_opening`, `binary_closing`, `binary_hit_or_miss`, `binary_propagation`, `binary_fill_holes`, `generate_binary_structure` and `iterate_structure`.

The only new kernel is the one corresponding to binary erosion. I was able to mostly reuse the existing utility `_filters_core._generate_nd_kernel`, but did have to add a couple of arguments to get it to work for this case. @coderforlife, please take a look if you have a chance.

All other binary morphological operations are built on top of `binary_erosion`. Some of the specific data arrays used in the tests were borrowed from SciPy's test suite.

The only feature in SciPy that is not implemented here is the case where `iterations > 1` and `brute_force=False`. SciPy uses a different algorithm in that case, but I have only implemented the brute force kernel here.

### CODE REVIEW

**Comment by grlee77** [comment]
I put some benchmarks for a (4096, 4096) binary image below. I only benchmarked binary erosion as all of the other functions also use it under the hood.

## Benchmark Script

```Python
import numpy as np
import cupy as cp
from cupyx.time import repeat
from scipy import ndimage as ndi
from cupyx.scipy import ndimage as cndi
shape = (4096, 4096)
x = np.random.randn(*shape) > 0.6
s = ndi.generate_binary_structure(x.ndim, 1)

# run 1 iteration on the CPU
perf = repeat(ndi.binary_erosion, (x, s), n_warmup=1, n_repeat=10)
print(perf)

# run 1 iteration on the GPU
xg = cp.asarray(x)
sg = cndi.generate_binary_structure(x.ndim, 1)
perf_gpu = repeat(cndi.binary_erosion, (xg, sg), n_warmup=20, n_repeat=100)
print(perf_gpu)

# compare with 5 iterations on the CPU
kwargs = dict(iterations=5, brute_force=True)
perf_cpu_5iter_bf = repeat(ndi.binary_erosion, (x, s), kwargs, n_warmup=1, n_repeat=10)
print(perf_cpu_5iter_bf)

kwargs = dict(iterations=5, brute_force=False)
perf_cpu_5iter = repeat(ndi.binary_erosion, (x, s), kwargs, n_warmup=1, n_repeat=10)
print(perf_cpu_5iter)

# compare with 5 iterations on the GPU (only brute_force=True is implemented)
xg = cp.asarray(x)
sg = cndi.generate_binary_structure(x.ndim, 1)
kwargs = dict(iterations=5, brute_force=True)
perf_gpu_5iter = repeat(cndi.binary_erosion, (xg, sg), kwargs, n_warmup=20, n_repeat=100)
print(perf_gpu_5iter)
```

## Benchmark Results

### single iteration CPU
```
binary_erosion      :    CPU:161958.266 us   +/-1280.388 (min:159108.258 / max:163741.605) us     GPU-0:161991.273 us   +/-1279.958 (min:159142.853 / max:163778.748) us
```

### single iteration GPU
```
binary_erosion      :    CPU:   57.767 us   +/-36.566 (min:   46.448 / max:  407.474) us     GPU-0: 1952.998 us   +/-81.631 (min: 1900.544 / max: 2227.200) us
```

### CPU, 5 iterations, brute_force=True
```
binary_erosion      :    CPU:492891.860 us   +/-3904.152 (min:486823.349 / max:499552.778) us     GPU-0:492924.872 us   +/-3903.662 (min:486854.309 / max:499583.923) us
```
### CPU, 5 iterations, brute_force=False
```
binary_erosion      :    CPU:263275.548 us   +/-8081.263 (min:257600.988 / max:281175.795) us     GPU-0:263310.513 us   +/-8081.219 (min:257634.888 / max:281208.618) us
```

### GPU, 5 iterations, brute_force=True
```
binary_erosion      :    CPU: 4045.870 us   +/-225.893 (min: 3803.620 / max: 4516.481) us     GPU-0: 4051.458 us   +/-226.087 (min: 3808.384 / max: 4531.968) us
```

**Comment by grlee77** [comment]
I think the call to `_center_is_true` has to synchronize as it checks if the central pixel of the structuring element is True or False.

For cases with `iterations > 1` there is also device synchronization because of the call: `changed = not (tmp_in == tmp_out).all()`. This check is used to detect if any of the values have changed since the prior iteration.

File: cupyx/scipy/ndimage/morphology.py [hunk 1]
```diff
@@ -1,13 +1,595 @@
...
+        rank(int): Number of dimensions of the array to which the structuring
+            element will be applied, as returned by `np.ndim`.
+        connectivity(int): `connectivity` determines which elements of the
+            output array belong to the structure, i.e., are considered as
+            neighbors of the central element. Elements up to a squared distance
+            of `connectivity` from the center are considered neighbors.
+            `connectivity` may range from 1 (no diagonal elements are
+            neighbors) to `rank` (all elements are neighbors).
+        on_cpu(bool, optional): If True, return a NumPy array rather than
+            transferring the structure to the GPU. (Default value = False)
```
Comment (asi1024) [inline cupyx/scipy/ndimage/morphology.py:133]: Why is this option needed?

File: cupyx/scipy/ndimage/morphology.py [hunk 2]
```diff
@@ -1,13 +1,595 @@
...
+            for example), to be dilated with itself.
+        iterations(int): The number of dilations performed on the structure
+            with itself.
+        origin(int or tuple of int, optional): If origin is None, only the
+            iterated structure is returned. If not, a tuple of the iterated
+            structure and the modified origin is returned.
+
+    Returns:
+        cupy.ndarray: A new structuring element obtained by dilating
+             `structure` (`iterations` - 1) times with itself.
```
Comment (asi1024) [inline cupyx/scipy/ndimage/morphology.py:96]: Please use double backticks.
```suggestion
             ``structure`` (``iterations`` - 1) times with itself.
```

File: cupyx/scipy/ndimage/morphology.py [hunk 3]
```diff
@@ -1,13 +1,595 @@
...
+
+    if not input.flags.c_contiguous:
+        # TODO: grlee77: is C-contiguity required?
+        input = cupy.ascontiguousarray(input)
+    if input.dtype.kind == 'c':
+        raise TypeError('Complex type not supported')
+    if structure is None:
+        structure = generate_binary_structure(input.ndim, 1)
+    else:
+        structure = cupy.asarray(structure, dtype=bool)
```
Comment (asi1024) [inline cupyx/scipy/ndimage/morphology.py:172]: Please use `astype` is the given `structure` is guaranteed to be `cupy.ndarray`.

File: cupyx/scipy/ndimage/morphology.py (hunk 1 above)
Comment (grlee77) [reply cupyx/scipy/ndimage/morphology.py:133]: Good catch. I meant to remove this option before making the PR. If a CPU result is desired, the user can always just call the SciPy function instead.

File: cupyx/scipy/ndimage/morphology.py (hunk 2 above)
Comment (grlee77) [reply cupyx/scipy/ndimage/morphology.py]: I have tried to use double backticks uniformly now.
Some existing functions in this file actually used **triple** backticks and I have also changed those to double backticks.

**Comment by jakirkham** [comment]
Jenkins, test this please.

**Comment by grlee77** [comment]
It seems very odd that only a single parameterized case out of 1728 failed for `BinaryErosionAndDilation`.

**Comment by jakirkham** [comment]
Sure, let's restart CI and see if it clears 🙂

Jenkins, test this please.

**Comment by jakirkham** [comment]
Would you be able to take another look @asi1024? 🙂

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
LGTM!

**Comment by jakirkham** [comment]
Thank you both! 😄

**Pull request merged**